test-data/unit/check-python38.test

```diff
@@ -240,4 +242,26 @@ def f(x: int = (c := 4)) -> int:
 
 reveal_type(c)  # E: Revealed type is 'builtins.int'
 
+def check_final() -> None:
+    x: Final = 3
+
+    if x := 4:  # E: Cannot assign to final name "x"
+        pass
+
+def check_binder(x: Optional[int], y: Optional[int]) -> None:
+    reveal_type(x)  # E: Revealed type is 'Union[builtins.int, None]'
+
+    (x := 1)
+    reveal_type(x)  # E: Revealed type is 'builtins.int'
+
+    if x or (y := 1):
+        reveal_type(y)  # E: Revealed type is 'Union[builtins.int, None]'
+
+def check_partial() -> None:
+    x = None
+    if bool() and (x := 2):
+        pass
```

Comment:
Maybe add a similar test where you first assign `(x := [])`, and then append something to `x`.